cudaungappedextender/include/claraparabricks/genomeworks/cudaungappedextender/cudaungappedextender.hpp

```diff
@@ -0,0 +1,90 @@
+/*
+* Copyright 2020 NVIDIA CORPORATION.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+#pragma once
+
+namespace claraparabricks
+{
+namespace genomeworks
+{
+namespace cudaungappedextender
+{
+
+struct Seed
+{
+    /// position of first sketch element in query_read_id_
+    position_in_read_t query_position_in_read;
+    /// position of second sketch element in target_read_id_
+    position_in_read_t target_position_in_read;
+};
+
+struct ScoredSegment
+{
+    /// Seed for the segment
+    Seed seed;
+    /// length of the segment
+    int32_t length;
+    /// score of the segment
+    int32_t score;
+};
+
+/// CUDA Ungapped Extension object
+class UngappedExtender
+{
+public:
+    /// \brief Constructor Prototype
+    UngappedExtender(int* h_sub_mat, int input_xdrop, bool input_noentropy,
+                     int gpu_id = 0, cudaStream_t stream = 0);
+
+    /// \brief Host pointer prototype for ungapped extension
+    ///
+    /// Takes values from host data structures,
+    /// copies them over to device,
+    /// launches async extension kernels on specified stream. Filters
+    /// segments on device based on input_hspthresh.
+    StatusType extend_async(const char* h_query, int32_t query_length,
+                            const char* h_target, int32_t target_length,
+                            int32_t score_threshold,
+                            std::vector<Seed>& h_seeds);
+
+    /// \brief Device pointer prototype for ungapped extension
+    ///
+    /// Memcopies to device memory are assumed to be done before this
+    /// function. Output array d_hsp_out must be pre-allocated on device.
+    /// Launches async extension kernel. Filters segments on device
+    /// based on input_hspthresh.
+    StatusType extend_async(const char* d_query, int32_t query_length,
+                            const char* d_target, int32_t target_length,
+                            int32_t score_threshold, Seed* d_seeds,
+                            int32_t num_seeds, ScoredSegment* d_hsp_out,
+                            int32_t* d_num_hsps);
```

Comment:
What exactly is `d_num_hsps`?

Reply:
It corresponds to the return value of UngappedExtend here: https://github.com/gsneha26/SegAlign/blob/GenomeWorks/src/seed_filter.cu#L1100
We would ideally like this function to return a completion status as it would be preferable to have all the ungapped extender work complete asynchronously. I'll have to check about the output compression to see if that decision can be made on the device without a CPU decision jump as it causes an unnecessary sync.

Reply:
1. The return value is of the type `uint32_t`, not an array.
2. So would you like to convert `int32_t* d_num_hsps` into the completion status array?
3. Could you elaborate on the part about the output compression?
3. Also, do you know what is going to be done about the `thrust` function calls in `UngappedExtend()`?

Reply:
1,2: Oh no it's a single int that is returned by the wrapper to indicate how many ScoredSegments to copy back to host. 
3: The decision to perform output compression can be done on the device by launching the kernels and exiting early with a condition check to see if num_hsps > 0 or by using dynamic parallelism (I'll have to check the perf impact of this as it can use up useful GPU resources). Also thrust kernels can be replaced by our own implementations or by using device_execution_policy (https://thrust.github.io/doc/structthrust_1_1device__execution__policy.html) for masking/unmasking the kernel launch (have to explore that too).

Reply:
To clarify, I'm aiming to make the device pointer version completely asynchronous.

Reply:
I think we should discuss this part about the post-processing in the device pointer and the corresponding private variables in the meeting/email thread.